### DISCUSSION TITLE
React强制刷新页面组件渲染的相关疑问

### DISCUSSION THREAD

**Discussion by nonzzz** [zxuqian/frontend-questions#10] [求助]
## 描述一下问题？
1.A组件通过react-router-dom向B传递参数。B接收参数进行渲染如果这时候刷新页面会短暂的出现组件的默认值。
2.useContext 和useRecuder模拟Redux中。发现如果useContext用f5刷新 状态会发生重置到初始值。

## 问题出现的场景、原因、截图和重现过程（适用于代码类）？
- 1 ![1606144723(1)](https://user-images.githubusercontent.com/52351095/99979721-60832b80-2de2-11eb-9e24-f91a5e8f7fc5.jpg)
这是正常路由切换展示的。但这时候我刷新页面,会有0.几s的
![a](https://user-images.githubusercontent.com/52351095/99979809-7bee3680-2de2-11eb-8297-3cd8abccb438.jpg)这样的效果。同时我点击圆形按钮进入C页面，C页面刷新也会发生入上图一样的问题。希望得到解答。

- 2 

如图所示。
![R4L7F9TK6QMMEH0`Q{7N2IB](https://user-images.githubusercontent.com/52351095/99980853-abea0980-2de3-11eb-81ff-40648fff0533.png)

` const editorItem = useContext(DataContext).toDoList[index] || []` 这一段我查阅过https://stackoverflow.com/的相关回答在react16以后我得通过这个方式获取到数据。比如我想拿到icon和colors我无法直接useContext(DataContext).toDoList[index].icon或者.colors 它会提示cant undefined it 的报错。想知道这个具体是因为什么原因呢。为什么要加这个|| 的条件呢。翻阅文档也没得到答案。

**Comment by zxuqian** [comment]
这俩问题都是正常的。单页应用在刷新的时候会重新从index.js入口开始执行。如果这个时候url是子页面路径，那么react-router 会在主程序渲染完成之后，再去跳转到子页面。

如果是全局状态也会重新初始化，没有手动备份的话以前的数据会丢失。因为单页应用的数据只针对当前加载的程序有效，是保存在内存中的，刷新相当于重新加载，也会清空上次加载的全部数据。

**Comment by zxuqian** [comment]
const editorItem = useContext(DataContext).toDoList[index] || []

如果不加|| []，假如todoList[index] 是undefined，那么访问undefined.icon肯定是报错，设置了默认值[]后，访问[].icon会返回undefined，因为属性不存在，但就不会报错了。
